src/Workspaces/Remote/ServiceHub/Services/CustomMessageHandler/RemoteCustomMessageHandlerService.cs

```diff
@@ -0,0 +1,252 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CustomMessageHandler;
+using Microsoft.CodeAnalysis.Remote.CustomMessageHandler;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Remote;
+
+internal sealed partial class RemoteCustomMessageHandlerService : BrokeredServiceBase, IRemoteCustomMessageHandlerService
+{
+#if !NETSTANDARD2_0
+    /// <summary>
+    /// Extensions assembly load contexts and loaded handlers, indexed by handler file path. The handlers are indexed by type name.
+    /// </summary>
+    private readonly ConcurrentDictionary<string, CustomMessageHandlerExtension> _extensions = new();
```

Comment:
I do still think that plugins from the same directory should have the same directory based ALC (we explicitly support that in analyzers for example).  But it isn't important for the initial prototype here.